x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,621 @@
+// needed to update the validator's power on the consumer chain at the end of the current block.
+func (k Keeper) DeleteKeyAssignmentReplacement(ctx sdk.Context, chainID string, providerAddr sdk.ConsAddress) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.KeyAssignmentReplacementsKey(chainID, providerAddr))
+}
+
+// AppendConsumerAddrsToPrune appends a consumer validator address to the list of consumer addresses
+// that can be pruned once the VSCMaturedPacket with vscID is received.
+//
+// The following invariant needs to hold:
+// For each consumer address cAddr in ValidatorByConsumerAddr,
+//   - either there exists a provider address pAddr in ValidatorConsumerPubKey,
+//     s.t. hash(ValidatorConsumerPubKey(pAddr)) = cAddr
+//   - or there exists a vscID in ConsumerAddrsToPrune s.t. cAddr in ConsumerAddrsToPrune(vscID)
+func (k Keeper) AppendConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscID uint64, consumerAddr sdk.ConsAddress) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ConsumerAddrsToPruneKey(chainID, vscID))
+	var consumerAddrsToPrune types.AddressList
+	if bz != nil {
+		err := consumerAddrsToPrune.Unmarshal(bz)
+		if err != nil {
+			panic(err)
+		}
+	}
+	consumerAddrsToPrune.Addresses = append(consumerAddrsToPrune.Addresses, consumerAddr)
+	bz, err := consumerAddrsToPrune.Marshal()
+	if err != nil {
+		panic(err)
+	}
+	store.Set(types.ConsumerAddrsToPruneKey(chainID, vscID), bz)
+}
+
+// GetConsumerAddrsToPrune returns the list of consumer addresses
+// that can be pruned once the VSCMaturedPacket with vscID is received
+func (k Keeper) GetConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscID uint64) [][]byte {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ConsumerAddrsToPruneKey(chainID, vscID))
+	if bz == nil {
+		return nil
+	}
+	var consumerAddrsToPrune types.AddressList
+	err := consumerAddrsToPrune.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
+	return consumerAddrsToPrune.Addresses
+}
+
+func (k Keeper) IterateConsumerAddrsToPrune(
+	ctx sdk.Context,
+	chainID string,
+	cb func(vscID uint64, consumerAddrsToPrune [][]byte) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iteratorPrefix := types.ChainIdWithLenKey(types.ConsumerAddrsToPruneBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, iteratorPrefix)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		_, vscID, err := types.ParseChainIdAndVscIdKey(types.ConsumerAddrsToPruneBytePrefix, iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+		var consumerAddrsToPrune types.AddressList
+		err = consumerAddrsToPrune.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(err)
+		}
+		stop := cb(vscID, consumerAddrsToPrune.Addresses)
+		if stop {
+			break
+		}
+	}
+}
+
+func (k Keeper) IterateAllConsumerAddrsToPrune(
+	ctx sdk.Context,
+	cb func(chainID string, vscID uint64, consumerAddrsToPrune [][]byte) (stop bool),
```

Comment:
ditto: `consumerAddrsToPrune` should be a concrete type or alias, not `[][]byte`